evm/src/assets/sharedComponents/AccessControl.sol

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.25;
+
+import { BytesParsing } from "wormhole-sdk/libraries/BytesParsing.sol";
+import "./ids.sol";
+
+//rationale for different roles (owner, admin):
+// * owner should be a mulit-sig / ultra cold wallet that is only activated in exceptional
+//     circumstances.
+// * admin should also be either a cold wallet or Admin contract. In either case,
+//     the expectation is that multiple, slightly less trustworthy parties than the owner will
+//     have access to it, lowering trust assumptions and increasing attack surface. Admins
+//     perform rare but not exceptional operations.
+
+struct AccessControlState {
+  address   owner; //puts owner address in eip1967 admin slot
+  address   pendingOwner;
+  address[] admins;
+  mapping(address => bool) isAdmin;
+}
+
+// we use the designated eip1967 admin storage slot:
+// keccak256("eip1967.proxy.admin") - 1
+bytes32 constant ACCESS_CONTROL_STORAGE_SLOT =
+  0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103;
+
+function accessControlState() pure returns (AccessControlState storage state) {
+  assembly ("memory-safe") { state.slot := ACCESS_CONTROL_STORAGE_SLOT }
+}
+
+error NotAuthorized();
+error InvalidAccessControlCommand(uint8 command);
+error InvalidAccessControlQuery(uint8 query);
+
+event OwnerUpdated(address oldAddress, address newAddress, uint256 timestamp);
+event AdminsUpdated(address newAddress, bool isAdmin, uint256 timestamp);
```

Comment:
Now that I see this event, I think the timestamp parameter is redundant :thinking:

Reply:
Is it? I added it in the swap layer because I thought it would be good to easily figure out a timeline of when certain changes were made, especially when in a panic (because realistically, you'll only look up this event if something weird with the permissions is going on).